src/main/java/org/jitsi/videobridge/cc/vp8/VP8AdaptiveTrackProjectionContext.java

```diff
@@ -125,330 +96,393 @@ public VP8AdaptiveTrackProjectionContext(
+                }
+            }
+        }
 
-        return nextVP8FrameProjection;
+        return frame.isAccepted() && frame.getProjection().accept(vp8Packet);
     }
 
     /**
-     * Given a frame (specified by the SSRC and the timestamp that are specified
-     * as arguments), find the highest sequence number we've received from
-     * that frame.
-     *
-     * @param ssrc the SSRC of the frame
-     * @param timestamp the timestamp of the frame
-     * @return the highest sequence number we've received from the frame that is
-     * specified by the SSRC and timestamp arguments.
+     * Create a projection for this frame.
      */
-    private int getMaxSequenceNumberOfFrame(long ssrc, long timestamp)
+    @NotNull
+    private VP8FrameProjection createProjection(@NotNull VP8Frame frame, @NotNull Vp8Packet initialPacket)
     {
-        Map<Long, Integer> frameToMaxSequenceNumberMap
-            = ssrcToFrameToMaxSequenceNumberMap.get(ssrc);
-
-        if (frameToMaxSequenceNumberMap == null)
+        if (frameIsNewSsrc(frame))
         {
-            return -1;
+            return createLayerSwitchProjection(frame, initialPacket);
         }
 
-        return frameToMaxSequenceNumberMap
-            .getOrDefault(timestamp, -1);
+        return createInLayerProjection(frame, initialPacket);
     }
 
     /**
-     * Upon arrival of an RTP packet of a video frame (specified by its SSRC,
-     * its timestamp and its sequence number that are specified as arguments),
-     * elects and store the highest sequence number we've received from that
-     * frame.
-     *
-     * @param ssrc the SSRC of the frame
-     * @param timestamp the timestamp of the frame
-     * @param sequenceNumber the sequence number of the RTP packet,
-     * potentially the highest sequence number that we've received from that
-     * frame.
+     * Create a projection for this frame. It is the first frame sent for a layer.
      */
-    private void maybeUpdateMaxSequenceNumberOfFrame(
-        long ssrc, long timestamp, int sequenceNumber)
+    @NotNull
+    private VP8FrameProjection createLayerSwitchProjection(@NotNull VP8Frame frame, @NotNull Vp8Packet initialPacket)
     {
-        Map<Long, Integer> frameToMaxSequenceNumberMap
-            = ssrcToFrameToMaxSequenceNumberMap
-            .computeIfAbsent(ssrc, k -> new LRUCache<>(5));
+        assert(frame.isKeyframe());
+        assert(initialPacket.isStartOfFrame());
+        long nowMs = System.currentTimeMillis();
 
-        if (frameToMaxSequenceNumberMap.containsKey(timestamp))
+        int projectedSeqGap = 1;
+
+        if (lastVP8FrameProjection.getVP8Frame() != null &&
+            !lastVP8FrameProjection.getVP8Frame().hasSeenEndOfFrame())
         {
-            int previousMaxSequenceNumber
-                = getMaxSequenceNumberOfFrame(ssrc, timestamp);
+            /* Leave a gap for the unfinished end of the previously routed frame. */
+            projectedSeqGap++;
```

Comment:
Do we know here that we only need to leave a space for one?  Or could it be more and we don't know so we just leave a gap of 1?

Reply:
It could be more than 1, but we close the old projection so that gap will never be filled in.  The gap is to make it clear that the old frame is incomplete.

Arguably this is unnecessary since the old code left frames incomplete without gaps, but I feel more comfortable making sure streams which are incomplete at the bitstream level are also incomplete at the RTP level.